<a href="https://colab.research.google.com/github/mk-felix/AIFFEL_quest_cr/blob/main/Python/quest02/avengers.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Avengers Vocab 만들기

Vocabulary -> 단어가 특정 정수와 매칭되어 있는 딕셔너리 형태의 단어장

Quest : 문장을 입력받아서 정수로 리턴해주는 함수 만들것

1. .txt *불러오기*
2. 모든 문자 소문자로 변환
3. 모든 기호 제거
4. 단어 단위 분리
5. 단어 빈도수 딕셔너리 형태 저장
6. 빈도순 내림차순 정렬 후, 인덱스 부여
7. 텍스트 input() 입력받고, 정수 return 함수

## 파일을 googld drive에 업로드 한 후, colab에 파일 접근 권한을 줍니다.

In [ ]:
from google.colab import drive
drive.mount('/Quest2-AvengersVocab')

Drive already mounted at /Quest2-AvengersVocab; to attempt to forcibly remount, call drive.mount("/Quest2-AvengersVocab", force_remount=True).


## 파일 접근합니다. 권한은 Read로 합니다. text 구문을 avengers 변수에 올립니다.

In [ ]:
f = open("/content/Quest2-AvengersVocab/Avengers.txt", "r")
avengers = f.read()

# 디버그 용도로 avengers 변수의 내용을 프린트해봅니다.
#print(avengers)

In [ ]:
# 디버그 용도로 위에서 파일 open 디스크립터였던 f 의 type을 확인합니다.
# print(type(f))

In [ ]:
# 디버그 용도로 avengers 변수의 타입을 확인합니다.
# print(type(avengers))

<class 'str'>


# 전처리 함수 preprocess(string)

1. 모든 문자를 소문자로 변환합니다.
2. 모든 특수문자 내지는 기호를 제거합니다.
3. 단어 단위로 분리합니다.
4. 분리된 단어리스트를 반환합니다.

In [ ]:
import re

def preprocess(param):

    # 소문자로 변경
    lowerParam = param.lower()
    # lower() 함수는 input 값을 소문자로 변경하는 것이 아니라 소문자로 변경한 string을 반환합니다.
    # 아래 lowerParam값과 param 값을 비교해 보면 그 차이를 알 수 있습니다.
    # print(lowerParam)
    # print(param)


    # 특수문자 내지 기호 제거

    text_rmv = re.sub('[-=+,#/\?:^.@*\"※~ㆍ!』‘|\(\)\[\]`\'…》\”\“\’·]', ' ', lowerParam)
    text_rmv = ' '.join(text_rmv.split())

    # 단어단위 분리
    # split 함수를 통해서 빈칸을 seperator로 하여 단어들을 리스트로 집어넣어 줍니다.
    text_word = text_rmv.split(sep=' ')


    # print(text_rmv)
    # print(text_word, type(text_word))


    # 분리된 단어 리스트를 반환합니다.
    return text_word




인덱싱 함수 indexing(text_word)

1. 분리된 단어 리스트를 입력받습니다.
2. 단어 빈도수 딕셔너리를 생성합니다.
3. 딕셔너리 각 요소를 내림차순으로 정렬합니다.
4. 인덱싱합니다.
5. 최종완성된 {단어 : 인덱스}의 딕셔너리를 반환합니다.

In [ ]:
'''
오퍼레이터 모듈과 관련하여서 https://zephyrus1111.tistory.com/278 참조
'''


import operator

def indexing(text_word):
    # 단어 빈도수 딕셔너리 생성
    tempDic = dict()
    for word in text_word:
        if word in tempDic:  # 이미 dictionary에 단어가 있는 경우
            tempDic[word] += 1
        else:                      # dictionary에 단어가 없는 경우
            tempDic[word] = 1

    #print(tempDic)

    # 딕셔너리 각 요소를 내림차순으로 정렬합니다.
    dic = dict()
    tempList = sorted(tempDic.items(), key = operator.itemgetter(1), reverse=True)
    dic = dict(tempList)

    # print(dic)
    # print(type(dic))

    # 인덱싱합니다
    counter = 0
    for i in enumerate(dic):
      dic[i[1]] = counter
      counter += 1

    print(dic)

    # 최종완성된 {단어 : 인덱스}의 딕셔너리를 반환합니다.
    return dic

    # indexedDict = dict()
    # for idx, val in enumerate(sorted_dict):
    #    indexedDict[idx] = val
    # print(indexedDict)

    # #최종 완성된 indexed Dictionary를 반환합니다.
    # return indexedDict

In [ ]:
a = preprocess(avengers)
indexedDictionary = indexing(a)


{'you': 0, 'the': 1, 'i': 2, 'a': 3, 'to': 4, 's': 5, 'it': 6, 'we': 7, 'that': 8, 't': 9, 'of': 10, 'is': 11, 'and': 12, 'he': 13, 'this': 14, 'on': 15, 'in': 16, 'not': 17, 'what': 18, 'me': 19, 'for': 20, 're': 21, 'have': 22, 'be': 23, 'no': 24, 'know': 25, 'are': 26, 'can': 27, 'do': 28, 'out': 29, 'm': 30, 'your': 31, 'was': 32, 'loki': 33, 'they': 34, 'there': 35, 'don': 36, 'need': 37, 'my': 38, 'get': 39, 'but': 40, 'all': 41, 'up': 42, 'got': 43, 'if': 44, 'here': 45, 'did': 46, 'just': 47, 'like': 48, 'tesseract': 49, 'an': 50, 've': 51, 'about': 52, 'his': 53, 'how': 54, 'stark': 55, 'with': 56, 'so': 57, 'will': 58, 'now': 59, 'think': 60, 'gonna': 61, 'him': 62, 'down': 63, 'them': 64, 'us': 65, 'from': 66, 'come': 67, 'll': 68, 'well': 69, 'didn': 70, 'as': 71, 'sir': 72, 'right': 73, 'put': 74, 'world': 75, 'where': 76, 'one': 77, 'why': 78, 'should': 79, 'who': 80, 'our': 81, 'has': 82, 'tell': 83, 'off': 84, 'people': 85, 'guy': 86, 'back': 87, 'then': 88, 'd': 89, 'w

# 사용자 입력을 받고 원하는 결과값을 얻어냅니다.

In [ ]:
def process():
    # 텍스트 input() 입력받고, 정수 return 함수
    input_text = input("단어 입력하세요: ")

    # input_text를 전처리하고 분리된 단어 리스트를 얻어냄.
    seperatedInput = preprocess(input_text)
    print(seperatedInput)


    output = list()
    for sepWord in seperatedInput:
      # sepWord가 indexedDictionary에 있는지 확인
      for k in indexedDictionary.keys():

        if sepWord == k:
          # print(indexedDictionary[k])
          output.append(indexedDictionary[k])

        else:
          pass

    print(output)




In [ ]:
process()

단어 입력하세요: I am a boy
['i', 'am', 'a', 'boy']
[2, 121, 3, 543]


# 회고

### 김민규(03)
1. list, dictionary 다루는 것이 좀 더 익숙해질 필요가 있겠습니다.
2. 데이터 처리 후에 검증을 어떻게 할 것인지에 대해 고민하게 되었습니다. 즉, 예를 든다면 문제에서 출력 예시로 든 경우
'I am a boy"에서 boy가 541이 나오지만 우리 팀에서 작성한 프로그램으로는 543이 출력됩니다. 과연 어떤 값이 맞는지 어떻게 검증할 것인가? 라는 문제를 생각하게 되었습니다.
3. 파이썬에 좀 더 애정을 가져야겠습니다.


### 고은비
1. 아직 리스트, 딕셔너리를 다루는 부분에 있어서 어떤 걸 사용해야 할 지 아리쏭한 것 같습니다. 더 익숙해질 필요성을 많이 느꼈습니다.
2. 예시답안과 동일하게 넣었으나, 값이 다른 건 txt 파일에서 사용된 문장이 조금 다른건지, 아니면 다른 요소의 작용이 있는지 궁금하지만, 해결할 방법은 모르겠습니다.